### PR TITLE
PR: Add automatic upload

### LINKED ISSUE #329: A-Cloud - Upload thumbnail image with packages
Request from Conda-Forge and also our Mosaic team - use case: they need to get an image uploaded with a package so that it can appear in Navigator as an icon (when the package is an application)

Should use the same syntax as notebook upload. If already specified in the meta.yml file, then just include that.

### PR DESCRIPTION
**Pull request by goanpeca**
Fixes #329

----

# Description
This PR automatically looks for the `icon` entry if defined inside the `app` section of the `recipe.json` file found inside the tar file after a conda build. This PR assumes that the icon (whatever name it has) is stored inside the conda recipe folder. By doing this, we avoid changing conda build logic.

# Actions
* [x] Implement pep8++ compliant code. (Optional for documentation updates)
* [x] Implement code that does not raise any flake8 errors. (Optional for documentation updates)
* [x] Add tests for the implemented code. (Optional depending if UI code)
* [x] Ensure all tests pass.
* [x] Ensure [coverage](https://coveralls.io/github/ContinuumIO/navigator) does not drop below the previous value.
* [x] Ensure at least one member of the team has reviewed the code.
* [ ] Update Changelog.
* [ ] Update [docs.continuum.io](https://docs.continuum.io/anaconda/navigator) (Optional if applies).

### CODE REVIEW

**Comment by goanpeca** [comment]
@dsludwig, I still need to add the test, but if you have some input...

**Comment by MulticastMatt** [comment]
@goanpeca - what's the latest here?

**Comment by goanpeca** [comment]
@srossross could you take a look at this?

I could write a more thorough test to actually make sure icons inside a package are parsed correctly.

File: binstar_client/__init__.py [hunk 1]
```diff
@@ -369,7 +369,8 @@ def remove_release(self, username, package_name, version):
         self._check_response(res, [201])
         return
 
-    def add_release(self, login, package_name, version, requirements, announce, description):
+    def add_release(self, login, package_name, version, requirements, announce,
+                    description, icon=None):
```
Comment (srossross) [inline binstar_client/__init__.py]: Can you check that the icon is a maximum size?

File: binstar_client/inspect_package/conda.py [hunk 2]
```diff
@@ -102,17 +124,22 @@ def inspect_conda_package(filename, fileobj, *args, **kwargs):
 
     package_data = {
                     'name': index.pop('name'),
+                    # Should this info be removed and moved to the release?
```
Comment (srossross) [inline binstar_client/inspect_package/conda.py]: Yes, I think that would be good

File: binstar_client/__init__.py (hunk 1 above)
Comment (goanpeca) [inline binstar_client/__init__.py]: That is done in another place (I used the upload thumbnail logic which incorporates that part)

Comment (srossross) [inline binstar_client/__init__.py]: OK, great. Can you add that info to the docstring?

Comment (goanpeca) [inline binstar_client/__init__.py]: :+1:

File: binstar_client/inspect_package/conda.py (hunk 2 above)
Comment (goanpeca) [inline binstar_client/inspect_package/conda.py]: Ok I will do that in a separate PR?

**Comment by srossross** [comment]
@goanpeca how do I test this? can you add an example package with an icon?

**Comment by goanpeca** [comment]
> @goanpeca how do I test this? can you add an example package with an icon?

**Comment by goanpeca** [comment]
@srossross added tests on a fake package that will be produced by a new version of conda-build.

**Comment by goanpeca** [comment]
@srossross I dont know why this fails on windows :-|

**Comment by MulticastMatt** [comment]
@srossross - @goanpeca believes he's done here unless you have any suggestions.  Please let us know.

**Comment by srossross** [comment]
@goanpeca can you add an icon to the conda package in example packages?

**Comment by goanpeca** [comment]
it is inside this onehttps://github.com/Anaconda-Platform/anaconda-client/blob/105e1c4c1224113b895a4fe53f53b1dbc5c61065/binstar_client/inspect_package/tests/data/test-app-package-icon-0.1-0.tar.bz2

what do you mean @srossross?

**Comment by srossross** [comment]
@goanpeca can you ad the recipe for that test-app-icon into the `example-packages` directory?

**Comment by goanpeca** [comment]
@srossross, @MulticastMatt added the recipe to the examples.

**Comment by srossross** [comment]
@goanpeca LGTM. thanks.

**Comment by srossross** [comment]
The upload works on my local install

**Pull request merged**